[아이작]미션3 - 체스판 초기화 구현

isaac56 opened the pull request:
안녕하세요 아이작입니다.
2021 코드스쿼드 체스만들기 미션 Step3 구현하였습니다.
부족한 코드지만 리뷰 잘 부탁드립니다~!

1. Board에서 모든 체스말을 관리할 수 있도록, Piece라는 abstract class를 만들었습니다.
2. Piece는 모든 체스말이 상속받아야 할 abstract class 입니다. 
 - 각 체스말은 color라는 공통 속성을 갖고 있고, 
 - Board에 출력하기 위한 representation 문자나 이동 범위 등 체스말 종류별로 다르게 구현해야하는 것도 있어서 abstract class로 구현했습니다.
3. 체스말 color 값은 상수에서 enum으로 변경하였으며, Piece class 내부에 Color enum으로 선언하였습니다.
4. Board 클래스에서는 Piece의 2차원 배열로 체스말을 관리하도록 했습니다.
5. Board 클래스 내에서 흰색말, 검은말의 목록을 따로 관리하도록 List<Piece> 변수를 각각 생성했습니다.

isaac56 committed:
Change access modifier from public to package-private in test code

test 코드의 public이었던 접근제어자를 package-private(default)로 수정하였습니다.

isaac56 committed:
Remove unused Main, MainTest class

사용하지않는 Main과 MainTest 클래스 삭제하였습니다.

isaac56 committed:
Change the package name in detail

패키지명을 보다 상세하게 수정하였습니다.
chess -> net.isaac.chess
pieces -> net.isaac.pieces

isaac56 committed:
Change PawnTest's method name in detail

PawnTest 클래스의 테스트 메소드명을 자세하게 수정하였습니다.

isaac56 committed:
Change all test code to use assertAll method

테스트 메소드 중 복수의 테스트를 진행하는 경우에 assertAll을 사용하도록 수정하였습니다.

isaac56 committed:
Change "pawns" from ArrayList to 2d array in Board class

1. Board class의 pawns 인스턴스 변수를 ArrayList에서 2차원 배열로 수정
2. Board class의 add와 findPawn 메소드를 2차원 배열에 맞게 수정
3. BoardTest class의 테스트 메소드 수정된 Board class에 맞게 수정

isaac56 committed:
Change Pawn's color from constant to PieceColor Enum

Pawn class에 상수로 정의되어있던 WHITE, BLACK의 색을 PieceColor라는 Enum으로 분리하였습니다.

isaac56 committed:
Create Piece abstract class

모든 체스말이 상속해야할 Piece abstract class 생성

isaac56 committed:
Modify Pawn class to implement Piece interface

Pawn 클래스가 Piece 추상클래스를 상속하도록 수정

isaac56 committed:
Change Board to handle Piece instead Pawn

Board가 Pawn이 아니라 Piece를 등록하고 관리하도록 수정

isaac56 committed:
Add initialize method in Board class

Board 클래스에 initialize 메소드 추가
initialize(): 2번째 행에 검은색 Pawn을 세팅하고 마지막에서 2번째 행에 하얀색 Pawn을 세팅

isaac56 committed:
Add print method in Board class

Board 의 체스말 상태를 출력하는 print 메소드 추가

isaac56 committed:
Rename method from getBoardCharacter to getRepresentation in  Piece class

step3 요구사항에 맞게 메소드명 수정

isaac56 committed:
Move getColor method from Pawn to Piece

상속받은 모든 클래스가 사용할 수 있게 getColor 메소드의 위치를 Pawn에서 Piece로 옮겼습니다.

isaac56 committed:
Modify initialize method to use add method in Board class

initialize 메소드 내부에서 add 메소드를 사용하도록 수정
(배열에 넣는 것 이외에 필요한 작업이 필요할 경우 add 메소드에서 모두 처리하기 위하여)

isaac56 committed:
Modify Board class to manage white and black separately

Board class 내에서 하얀 체스말과 검은 체스말을 따로 관리하도록 하였습니다.
1. whitePieces, blackPieces 추가 (List<Piece>)
2. add 메소드 내에서 piece 색을 확인하여 알맞은 리스트에 추가

isaac56 committed:
Add getWhitePiecesResult, getBlackPiecesResult method in Board class

각 색깔 별로, board에 등록된 Piece의 Representation 문자를 붙여서 String으로 반환하는 메소드 추가
getWhitePiecesResult(): 하얀말 목록을 String으로 반환
getBlackPiecesResult(): 검은말 목록을 String으로 반환

isaac56 committed:
Add initializeTest method in BoardTest

Board의 initialize 메소드를 테스트하는 코드 추가

isaac56 committed:
Add getRepresentation method in Board class

Board 현황을 출력하기 위한 문자열을 반환하는 getRepresentation 메소드를 추가하였습니다.

isaac56 committed:
Create Main class that starts and ends a chess game.

사용자 입력을 받아 체스 게임을 시작하고 종료하는 Main 클래스를 생성하였습니다.

isaac56 committed:
Move PieceColor enum into Piece abstract class

PieceColor enum은 Piece abstract class에 종속적인 관계같아서,
Piece 클래스 내부로 이동하고 Color로 Rename하였습니다.

isaac56 committed:
Make color constructor in Piece

Piece 클래스에 Color를 설정하는 생성자를 만들었고, 기존의 Pawn 클래스 생성자는 부모 생성자를 사용하도록 했습니다.

sonypark commented on src/main/java/net/isaac/Main.java:
```diff
@@ -0,0 +1,31 @@
+import java.io.InputStreamReader;
+
+public class Main {
+    public static void main(String[] args){
+        String cmd;
+        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+        try{
+            while(!(cmd = reader.readLine()).equals("end")){
+                if(cmd.equals("start"))
+                    startChess();
+            }
+        }
+        catch (IOException e) {
+            e.printStackTrace();
+        }
+        System.out.println("Game end.");
+    }
```
Scanner는 생성 후 close를 해주는걸 권장합니다. 왜 그럴까요? 이유를 알았다면 어떻게 close 할 수 있을까요? try catch를 사용하셨으니, [try with resources](https://www.baeldung.com/java-try-with-resources)에 대해 공부해보세요~

isaac56 replied:
Scanner 뿐만 아니라 외부 리소스는 사용이 끝나면 반환해야 하군요.
다른 프로세스 또는 쓰레드에서 리소스를 사용할 수록 있도록요.
단일 콘솔 프로그램의 경우 프로그램이 종료되면 자동으로 리소스를 반환하지만, 서버에서 close를 안하고 있다면 계속해서 리소스를 갖고 있을테니까요.

try with resources를 사용하면 블럭이 끝났을때 자동으로 리소스를 반환해주니 편리한 구문같습니다.
좋은 정보 감사합니다~!

sonypark commented on src/main/java/net/isaac/Main.java:
```diff
@@ -0,0 +1,31 @@
+
+public class Main {
+    public static void main(String[] args){
+        String cmd;
```
네이밍은 되도록 축약하지 않고 써주세요

sonypark commented on src/main/java/net/isaac/Main.java:
```diff
@@ -0,0 +1,31 @@
+        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+        try{
+            while(!(cmd = reader.readLine()).equals("end")){
+                if(cmd.equals("start"))
```
변하지 않는 값은 가독성과 유지보수를 위해 자바 상수를 사용해보세요

sonypark commented on src/main/java/net/isaac/chess/Board.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    public Piece findPiece(int rowIdx, int colIdx){
+        if(!isValid(rowIdx) || !isValid(colIdx))
+            return null;
+
+        return pieces[rowIdx][colIdx];
```
if 문에는 중괄호를 꼭 기입해주세요

sonypark commented on src/main/java/net/isaac/chess/Board.java:
```diff
@@ -0,0 +1,85 @@
+
+        pieces[rowIdx][colIdx] = piece;
+
+        if(piece.getColor() == Piece.Color.BLACK)
+            blackPieces.add(piece);
+        else if(piece.getColor() == Piece.Color.WHITE)
+            whitePieces.add(piece);
+
```
else if 사용을 지양해주세요

isaac56 replied:
else if를 지양한다면 위 코드는 switch로 대체하는게 최선일 것 같습니다.
(if 블럭의 내용이 리턴이 아니기 때문에 [빠른 리턴을 통한 else 구문 제거](https://xly.kr/posts/54)는 힘들 것 같습니다.)
가독성을 위해서 else if 지양하라는 말씀이신가요?

sonypark replied:
음 다시 보니 이 부분은 단순한 구조라 switch로 대체 하지 않고 else if로 가도 괜찮을 것 같네요..!
되도록이면 else if는 사용하지 말자라는 주의라 위와 같이 리뷰를 드렸네요😅
[호눅스의 말씀](https://github.com/codesquad-members-2021/java-chess/pull/98#issuecomment-780485071)처럼 제 말을 꼭 안 따르셔도 됩니다.

isaac56 replied:
옙ㅎㅎ 사실 else if를 생각 없이 사용했었는데 덕분에 좀 더 고민하는 계기가 생겼습니다.
감사합니다~!

sonypark commented on src/main/java/net/isaac/chess/Board.java:
```diff
@@ -0,0 +1,85 @@
+        return pieces[rowIdx][colIdx];
+    }
+
+    private boolean isValid(int location){
+        if(location < 0 || MAX_SIZE <= location)
+            return false;
+        else
+            return true;
+    }
```
```suggestion
private boolean isValid(int location){
        return location >= 0 && MAX_SIZE > location;
    }
```
좀 더 간단히 표현 가능할 수도 있겠네요

sonypark commented on src/main/java/net/isaac/chess/Board.java:
```diff
@@ -0,0 +1,85 @@
+                else
+                    builder.append(pieces[i][j].getRepresentation());
+            }
+            builder.append(System.lineSeparator());
```
`lineSeparator()` 👍

sonypark commented on src/main/java/net/isaac/chess/Board.java:
```diff
@@ -0,0 +1,85 @@
+public class Board {
+    private static final int MAX_SIZE = 8;
+    private static final int MAX_IDX = MAX_SIZE - 1;
+
+    private Piece[][] pieces = new Piece[MAX_SIZE][MAX_SIZE];
```
2차원 배열로도 구현 가능하지만, 권장하지는 않습니다.
2차원 배열을 사용하면 코드의 가독성이 현저히 낮아지는데요, 다른 방법으로 구현할수 없는지도 고민해보세요~
당장 적용이 어렵다면 미션을 진행하며 천천히 고민하고 적용해도 좋습니다.

isaac56 replied:
체스 게임을 진행하면서 보드의 크기가 변하지 않기 때문에 2차원 배열을 사용했습니다.
체스는 2차원 좌표를 기반으로 말을 이동하기 때문에, 2차원 배열 사용 시 가독성이 오히려 높아진다고 생각했거든요.

제 수준으로는 2차원 좌표상에서 말을 관리하려면 `List<List<Piece>>` 또는 `Piece[][]` 밖에 떠오르지 않았습니다.
`List<List<Piece>>`를 사용하는 것보다는 `Piece[][]`를 사용하는게 더 간단하다고 생각했습니다.

혹시, 제가 리뷰어님께서 얘기하신 '다른 방법'과 전혀 다른 방향으로 생각하고 있는 걸까요..? ㅠㅠ

sonypark replied:
아닙니다. 이차원 배열로 구현하셔도 괜찮습니다. 말씀하신대로 `List<List<Piece>>`형태로도 가능합니다. List와 같은 자바의 컬렉션을 사용하면 컬렉션 메서드인 add(), size() 등을 사용할 수 있는 장점이 있습니다. 또 다른 방식으로는 Map 과 같은 자료구조로도 구현할 수 있겠네요. 구현 방법은 다양합니다. 이차원 배열이 가독성이 높고 더 구현하기 편하다고 느끼신다면 사용하셔도 좋습니다!

sonypark commented on src/main/java/net/isaac/pieces/Piece.java:
```diff
@@ -0,0 +1,21 @@
+        this.color = color;
+    }
+
+    //Board 출력을 위한 문자 반환
+    public abstract char getRepresentation();
+
```
representation을 표현하는 로직을 추상 메서드로 만드는것이 좋은 선택일까요?
representation은 Piece마다 갖고 있는 특성일 것 같은데 어떻게 하면 좋을까요?

isaac56 replied:
representation은 모든 Piece가 갖고 있는 특성이기 때문에 멤버 변수로 갖고 있어야 할 것 같습니다.
대신 상속받는 각 체스 말이 `abstract char getRepresentation(Color color)`라는 추상 메소드를 
Override하여 color에 따른 representation을 설정할 수 있도록 하려고 합니다.

```java
public abstract class Piece {
    Color color;
    char representation;

    public Piece(Color color){
        this.color = color;
        this.representation = getRepresentation(color);
    }

    abstract char getRepresentation(Color color);

    //Board 출력을 위한 문자 반환
    public char getRepresentation(){
        return representation;
    };

    public Color getColor() {
        return color;
    }

    public enum Color {
        BLACK,
        WHITE
    }
}
```

sonypark commented on src/main/java/net/isaac/pieces/Pawn.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+
+    public char getRepresentation(){
+        if(this.color == Color.BLACK)
+            return 'P';
+        else if(this.color == Color.WHITE)
+            return 'p';
+        else
+            return '?';
```
representation을 이렇게 하드코딩으로 표현해주는게 적절한지 고민해보세요🧐

isaac56 replied:
Piece 클래스에 상수로 각 체스말의 representation을 선언한 후에 그 상수를 사용하려고 합니다.
감사합니다~!

sonypark commented on src/main/java/net/isaac/chess/Board.java:
```diff
@@ -0,0 +1,85 @@
+        return builder.toString();
+    }
+    
+    public String getWhitePiecesResult(){
+        return this.whitePieces.stream().
+                map(x->String.valueOf(x.getRepresentation())).
+                reduce("",(a,b) -> a+b);
+    }
+
+    public String getBlackPiecesResult(){
+        return this.blackPieces.stream().
+                map(x->String.valueOf(x.getRepresentation())).
+                reduce("",(a,b) -> a+b);
+    }
```
이 메서드는 개선할 부분이 많아 보이는데요, representaion을 표현하는 로직을 개선하면 이 부분도 자연스럽게 변경될 것 같네요!

isaac56 replied:
"보드에 존재하는 white piece의 representation 문자를 String으로 묶어 반환한다"는 메소드의 목적 자체가 개선할 점인가요?
아니면 그런 메소드를 만들기 위한 구현 방법이 개선할 점인가요?

representation을 표현하는 로직이 고쳐진다고 해도, 결국 Piece 클래스 외부에서는 getter로 속성을 얻어야 할 것 같습니다.
그러면 로직이 개선되어도 위 코드에 영향이 가진 않을 것 같거든요.

제가 부족하여 리뷰를 잘 알아듣지 못했는데 좀 더 자세히 설명 부탁드려도 될까요..?

sonypark replied:
메서드 내부의 로직이 개선될 필요가 있다는 말씀이었습니다:)
문자열을 연결하는 로직이라면 reduce 가 아닌 `Collectors.joining()` 사용하는 방법도 있습니다. [참고](https://tourspace.tistory.com/8)
그리고 결국 문자열로 표시될 representation이라면 char로 선언해서 다시 `String.valueOf()`로 형변환 하는것보다 처음부터 String 자료형을 사용하는 건 어떨까요?

isaac56 replied:
Collectors.joining() <- 좋은 정보 감사합니다 ㅎㅎ

char 자료형을 사용한 이유는, 개발자가 실수로 representation 값에 여러 문자를 넣는걸 방지하기 위해서였습니다.
체스판에 표시하는 representation 문자는 1개여야 한다고 생각했어요.(출력 형식을 맞추기 위해서)
```java
representation = "AF"
```
이런식으로 들어가는 걸 방지하기 위해서 char 자료형으로 선언하였어요.

체스판에 존재하는 말의 개수가 수백만개가 아닌 이상 String.valueOf()로 형변환하는 오버헤드가 크지 않기 때문에,
보다 엄격하게 자료형을 제한해서 실수를 방지하는게 낫다고 생각했습니다.

sonypark reviewed (changes requested):
안녕하세요 아이작, 리뷰어 소니 입니다.😊
3단계 미션 구현하시느라 수고 많으셨습니다.
2차원 배열, 추상 메서드 사용 등 구현을 위해 고민한 흔적들이 보이네요.
그런데 이 부분이 원하는 기능을 구현하기 위해 적절한 선택이었는지는 고민해봐야할 것 같아요.
더 좋은 방법은 없는지, 이렇게 구현했을 때의 문제는 무엇인지 생각해보시면 좋을 것 같습니다.
코멘트 확인하시고 궁금한 점은 질문 남겨주세요~

isaac56 committed:
Modify Main class to use try-with-resource block

코드 리뷰 내용 반영
1. 사용이 끝난 리소스를 자동으로 반환하도록 try-with-resource 구문 사용
2. 변수 네이밍은 되도록 축약하지 않도록 cmd를 command로 수정

isaac56 committed:
Replace "start","end" with constants in Main class

코드 리뷰 반영
Main 클래스에서 게임 시작 종료 명령을 판단하기 위해 사용하던 "start", "end" 문자열을 상수로 변경하였습니다.

isaac56 committed:
Simplify isValid method in Board class

Board class의 isValid 메소드 간결화

Co-authored-by: Kong Son Park <sonypark0204@gmail.com>

isaac56 committed:
Modify how to get representation character from Piece class

representation 문자를 멤버변수로 추가하고, Piece 생성할 때 설정해주도록 수정하였습니다.
기존의 getRepresentation()은 getter입니다.

abstract char getRepresenataion(Color color) 추상메소드:
- Piece를 상속받는 각 체스말이 Override하여 색깔에 따른 자신의 representation을 설정하여야 합니다.

isaac56 committed:
Change 'else if' to 'switch' in Board class

코드 리뷰 반영
else if 구문 지양하기 위하여 switch 문으로 변경하였습니다.

isaac56 committed:
Change 'else if' to 'early return' in Pawn class

코드 리뷰 반영
else if 구문을 지양하기 위하여 early return 방식으로 수정하였습니다.

isaac56 committed:
Add representation constant to avoid hard coding in Piece

하드 코딩을 피하기 위해, Piece 클래스에 representation의 상수를 추가하였습니다.
1. Piece class에 representation 상수 추가
2. Pawn class의 getRepresentation(Color color)에서 representation 상수 사용

sonypark commented on src/main/java/net/isaac/pieces/Piece.java:
```diff
@@ -0,0 +1,31 @@
+package net.isaac.pieces;
+
+public abstract class Piece {
+    public static final char WHITE_PAWN_REPRESENTATION = 'p';
+    public static final char BLACK_PAWN_REPRESENTATION = 'P';
+    public static final char UNKNOWN_REPRESENTATION = '?';
```
지금과 같은 구조라면 체스말의 종류가 늘어나는 만큼 자바 상수도 계속 추가될 것 같은데 맞나요?

isaac56 replied:
옙! 체스말의 종류에 따라 representation은 달라집니다.
체스말 종류마다 representation을 적어도 한번은 지정해놔야 하기 때문에, 체스말 종류만큼 상수가 생겨야 할 것 같습니다.
다만 WHITE, BLACK까진 나누지 않고 검은색일땐 대문자, 흰색일땐 대문자로 바꾸는 로직을 추가하면 될 것 같습니다.

sonypark commented on src/main/java/net/isaac/pieces/Piece.java:
```diff
@@ -0,0 +1,31 @@
+    public static final char BLACK_PAWN_REPRESENTATION = 'P';
+    public static final char UNKNOWN_REPRESENTATION = '?';
+
+    Color color;
+    char representation;
```
접근제어자를 default로 하신 이유가 있을까요?

isaac56 replied:
Piece 상속받는 체스말 클래스를, Pieces 패키지안에 만들게하려는 의도입니다.
protected로 할 경우 Pieces 패키지 밖에서도 상속받아 사용할 수 있기 때문에 default로 하였습니다.

sonypark replied:
`Piece 상속받는 체스말 클래스를, Pieces 패키지안에 만들게하려는 의도`
이 부분이 잘 이해되지 않습니다.
Piece가 갖고 있는 필드를 외부에서 사용하지 않고 Piece 클래스 내에서 사용하도록 로직을 구현하면 되지 않을까요? 

예를 들어, color 비교도 상속받는 클래스까지 안가더라도 Piece 클래스 안에서도 할 수 있을 것 같은데
```java
public boolean isSameColor(Color color) {
        return this.color == color;
    }
```
필드의 접근 제어자는 최소한으로 해주는게 좋습니다. 자식 클래스에서 부모 클래스의 필드를 가져다 쓰는게 좋은 설계인지도 생각해보셨으면 합니다.

isaac56 replied:
생각해보니 color나 representation 변수 모두 상속받는 클래스에서 사용할 필요가 없었네요!
왜인지는 모르게 상속받는 클래스에서 사용해야 한다고 생각하고 있었어요! 
좋은 지적 감사합니다ㅎㅎ

sonypark commented on src/main/java/net/isaac/pieces/Piece.java:
```diff
@@ -0,0 +1,31 @@
+    Color color;
+    char representation;
+
+    public Piece(Color color){
+        this.color = color;
+        this.representation = getRepresentation(color);
+    }
```
필드로 representation을 두셨는데, 생성자를 통해 초기화 하는 방식은 어떨까요?

isaac56 replied:
생성자로 representation을 받아서 초기화하면, 생성자를 호출하는 코드에서 
```java
new Piece(Color.WHITE, '&')
```
와 같이 잘못된 문자를 넣을 수 있기 때문에, 추상메소드를 오버라이드하여 자체 로직으로 초기화하도록 했습니다.

다만, Piece의 정적 팩토리 메소드로 각 체스말을 생성하게 한다면 위에 말한 문제는 없을 것 같아요!
정적 팩토리 메소드를 사용할 지, 지금처럼 진행할지는 다음 스텝을 진행하면서 고민해보려고 합니다:)

sonypark replied:
네 맞습니다. 잘못된 문자가 들어올 수 있는 경우가 생기겠네요. 그렇다면 체스말의 representation만 들어오도록 타입을 강제하는 방법은 없을까요? char라는 기본 자료형이 아니라 자기가 원하는 타입만 들어오도록 강제하도록 하고 싶다면 어떻게 해야할까요?

isaac56 replied:
제 생각은 타입을 강제한다기 보다 값을 강제해야 할 것 같습니다.
Color처럼 Representation이라는 enum을 만들어도, Pawn 생성자에 Representation.QUEEN과 같이 이상한 값을 넣을 수 있다고 생각해서요.

Piece 클래스 내부에 아래와 같은 static method를 정의하고,
```java
public Piece getPawn(Color color) { return new Pawn(color, Piece.PAWN_REPRESENTATION); }
```

Pawn 생성자의 접근제어자를 default로 제한하여, Piece.getPawn 메소드를 통해서만 Pawn의 인스턴스를 생성할 수 있도록 하려고 합니다.

sonypark replied:
네 이번 단계에서는 말씀하신대로 진행해주세요~

++ 참고로 타입을 사용하더라도 각 체스말의 생성자에서 자신의 representation을 강제할 수 있습니다.

```java
public Pawn(Color color) {
        super(color, Representation.PAWN);
    }
```

sonypark commented on src/main/java/net/isaac/pieces/Pawn.java:
```diff
@@ -0,0 +1,21 @@
+        super(color);
+    }
+
+    @Override
+    public char getRepresentation(Color color){
+        if(this.color == Color.BLACK)
+            return Piece.BLACK_PAWN_REPRESENTATION;
+        if(this.color == Color.WHITE)
+            return Piece.WHITE_PAWN_REPRESENTATION;
+
+        return Piece.UNKNOWN_REPRESENTATION;
+    }
```
representation은 결국 화면에 표시하기 위한 부분입니다. 색상에 따라 출력이 달라지는 로직을 모든 Piece마다 각각 중복해서 갖고 있는게 좋은 구조일까라는 생각이 드네요. Piece의 representation을 결정짓는 부분, 즉 흰색이면 소문자, 검은색이면 대문자로 표시하는 로직은 어디에서 해주는게 좋을까요? 한 번 고민해보시고, 생각이 잘 나지 않는다면 추후에 변경해보셔도 좋습니다. 위와 같은 구조를 이어간다면 어떤 문제가 있을지 생각해보시고, 문제점이 보인다면 개선해주세요. 아직 잘 모르겠거나 큰 문제가 되지 않을 것 같다면 그대로 진행해주셔도 됩니다.

++ if 문에는 중괄호를 생갹하지 않고 사용해주세요

isaac56 replied:
상속받는 체스말마다 위와 비슷한 코드가 중복되는 문제가 생길 것 같습니다.

저 추상메소드를 삭제하고 Piece에 정적 메소드를 만들어 Pawn을 생성하게 하려고 합니다.
그리고 Piece의 representation getter에서, Color에 따라 대문자 또는 소문자를 반환하도록 하려고 합니다.

if 문에는 항상 중괄호를 넣는 습관을 가져야겠네요 ㅠㅠ 감사합니다!

sonypark commented on src/main/java/net/isaac/chess/Board.java:
```diff
@@ -0,0 +1,86 @@
+            add(new Pawn(Piece.Color.WHITE), MAX_IDX-1, i);
+        }
+    }
+
+    public void print(){
+        System.out.println(getRepresentation());
+    }
+
+    private String getRepresentation(){
+        StringBuilder builder = new StringBuilder();
+        for(int i = 0; i < MAX_SIZE; i++) {
+            for(int j = 0; j < MAX_SIZE; j++) {
+                if(pieces[i][j] == null)
+                    builder.append('.');
+                else
+                    builder.append(pieces[i][j].getRepresentation());
+            }
+            builder.append(System.lineSeparator());
```
출력 로직을 Board 클래스에서 갖고 있는게 맞는지도 생각해봐주세요~
예를 들어, 출력을 담당하는 객체에게 체스판을 넘겨주면 해당 객체가 체스판을 출력하는 흐름은 어떨까요?
이 부분은 당장 고치지 않으셔도 되고, 미션 진행하면서 필요성을 느낄 때 분리해주시면 됩니다.

sonypark commented on src/main/java/net/isaac/chess/Board.java:
```diff
@@ -0,0 +1,86 @@
+                reduce("",(a,b) -> a+b);
+    }
+
+    public boolean add(Piece piece, int rowIdx, int colIdx){
+        if(!isValid(rowIdx) || !isValid(colIdx))
+            return false;
```
add 메서드의 반환 값을 boolean으로 하신 이유는 무엇인가요??

isaac56 replied:
rowIdx나 colIdx가 잘못된 값으로 들어와, add에 실패했을 때를 알기 위해서 boolean을 반환했습니다.
add 실패하면 false 반환, 성공하면 true 반환하도록요.

sonypark replied:
boolean을 리턴해서 add의 성공여부를 판단하려는 의도였다면, 그 의도에 맞게 사용되는 부분이 있으면 좋을 것 같아요. 테스트 케이스도 추가해주세요.

개인적으로 하나의 메서드는 하나의 역할만 하도록 최대한 분리하는걸 선호합니다. 하나의 메서드가 여러 역할을 하게되면 리팩토링이나 테스트를 하기 어려워지기 때문입니다. add 메서드의 역할은 체스판에 체스말을 추가하는 것 입니다. 잘 추가가 되었는지는 add 메서드에서 하는게 아니라 조회를 통해 확인하는게 자연스러운 것 같습니다. 

리턴 타입과는 별개로 현재 rowIdx와 colIdx가 체스판 범위를 벗어났을 때 아무 예외 메시지 없는데요. 사용자에게 예외 메시지를 통해 왜 add가 안 되었는지 알려줄 필요가 있다고 생각합니다.

isaac56 replied:
오늘 호눅스 수업에도 들었지만, 제가 '지금은 필요없지만 언젠가 필요하지 않을까?'라는 생각으로 코딩하고 있었던 것 같습니다.
나중에 필요할 때 refactoring을 해도 되는데, 안좋은 습관이었던 것 같아요.
소니 말씀대로 한 메서드가 하나의 역할만 하는게 더 자연스러운 것 같습니다. 감사합니다!

sonypark commented on src/main/java/net/isaac/Main.java:
```diff
@@ -0,0 +1,32 @@
+    public static final String END = "end";
+    public static void main(String[] args){
+        String command;
+        try(BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))){
+            while(!(command = reader.readLine()).equals(END)){
+                if(command.equals(START))
+                    startChess();
+            }
+        }
+        catch (IOException e) {
+            e.printStackTrace();
+        }
+        System.out.println("Game end.");
```
try with resource 👍

sonypark reviewed (changes requested):
질문 주신 부분에 대해 코멘트 남겼습니다.
말씀 드린 내용 중 이해가 안 되는 부분 있으면 질문 남겨주세요~
아이작의 생각도 가감없이 남겨주시면 감사하겠습니다!

isaac56 committed:
Change add method's return type from boolean to void in Board class

1. add 메소드가 '체스말을 추가한다'는 역할만 하도록 반환타입을 void로 변경하였습니다.
2. 추가하려는 위치가 범위를 벗어났을 경우 예외메세지 출력하도록 하였습니다.

isaac56 committed:
Remove basic constructor from Pawn class

Pawn 클래스에서 기본생성자 삭제

isaac56 committed:
Remove getRepresentation abstract method from Piece class

필요없는 getRepresentation 추상메소드 제거
(생성자에서 representation을 받아 초기화하도록 할 예정)

isaac56 committed:
Change Piece constructor to initialize representation using parameter

Piece 생성자에서 representation 을 파라미터로 받아 초기화하도록 수정

isaac56 committed:
Change color, representation access modifier to private in Piece class

Piece 클래스의 color와 representation 접근제어자를 private으로 변경

isaac56 committed:
Force to get new instance of Pawn using static factory method in Piece class.

1. Piece의 getPawn 정적 메소드를 통해서 Pawn의 Instance를 생성하도록 하였습니다.
2. Pawn의 생성자의 접근제어자를 default로 하여 Piece의 getPawn을 통해서만 생성할 수 있도록 하였습니다.

isaac56 committed:
Chagne getPawn method's return type from Piece to Pawn in Piece class

getPawn 메소드의 리턴타입을 Pawn으로 변경

honux77 commented:
@sonypark 제가 머지할게요. 수고하셨습니다.
